lucid/main_window.py

```diff
@@ -163,6 +307,572 @@ def wrapper(*args, **kwargs):
+                        callback=cell.click,
+                    )
+
+
+def _raise_display(display):
+    logger.debug('Bringing %s to the front', display)
+    display.raise_()
+    display.activateWindow()
+
+
+def _thread_screens_search(callback, *, general_search, category_search,
+                           threshold):
+    'Search open screens for the given text, running callback on each result'
+    if not general_search:
+        return
+
+    main = LucidMainWindow.get_instance()
+    for display in main.findChildren(typhon.TyphonDeviceDisplay):
+        ratio = max(utils.fuzzy_match(display.device_name, text,
+                                      threshold=threshold)
+                    for text in general_search)
+        if ratio > threshold:
+            callback(
+                source='screens',
+                rank=ratio,
+                name=display.device_name,
+                item=display,
+                reason='device',
+                callback=lambda disp=display: _raise_display(disp),
+            )
+
+    for suite in main.findChildren(typhon.TyphonSuite):
+        suite_parent = suite.parent()
+        if not hasattr(suite_parent, 'title'):
+            continue
+
+        ratio = max(utils.fuzzy_match(suite_parent.title, text,
+                                      threshold=threshold)
+                    for text in general_search)
+        if ratio > threshold:
+            callback(
+                source='screens',
+                rank=ratio,
+                name=suite_parent.title,
+                item=suite,
+                reason='suite',
+                callback=lambda disp=suite: _raise_display(disp),
+            )
+
+
+def _happi_dict_to_display(d):
+    name = d['name']
+
+    @LucidMainWindow.in_dock(title=f'[happi] {name}')
+    def wrapped():
+        client = utils.get_happi_client()
+        # dictionary -> device container -> device -> display
+        happi_device = client.find_device(**d)
+        device = happi.loader.from_container(happi_device)
+        return utils.display_for_device(device)
+
+    wrapped()
+
+
+def _thread_happi_search(callback, *, general_search, category_search,
+                         threshold):
+    'Search happi for the given text, running callback on each result'
+    global _HAPPI_CACHE
+    if _HAPPI_CACHE is None:
+        # TODO: re-read happi after a certain interval?
+        client = utils.get_happi_client()
+        _HAPPI_CACHE = list(client.search(as_dict=True))
+
+    for item in _HAPPI_CACHE:
+        item_results = []
+        for key, text in category_search:
+            value = item.get(key)
+            if value is not None:
+                ratio = utils.fuzzy_match(text, str(value),
+                                          threshold=threshold)
+                item_results.append((ratio, f'{key}: {value}'))
+
+        for text in general_search:
+            for key in ['name', 'prefix', 'stand']:
```

Comment:
Can we extract the keys from here and into a constant for easier update in the future if we decide to expand/modify?

Reply:
Done